src/http.rs

```diff
@@ -72,31 +85,298 @@ impl HttpError {
 pub type HttpClientFuture = <HttpClient as Service<http::Request<Body>>>::Future;
 type HttpProxyConnector = ProxyConnector<HttpsConnector<HttpConnector>>;
 
+#[async_trait]
+trait AuthExtension<B>: Send + Sync
+where
+    B: fmt::Debug + HttpBody + Send + 'static,
+    B::Data: Send,
+    B::Error: Into<crate::Error> + Send,
+{
+    async fn modify_request(&self, req: &mut Request<B>) -> Result<(), vector_lib::Error>;
+}
+
+#[derive(Clone)]
+struct OAuth2Extension {
+    token_endpoint: String,
+    client_id: String,
+    client_secret: Option<SensitiveString>,
+    grace_period: u32,
+    client: Client<HttpProxyConnector, Body>,
+    token: Arc<Mutex<Option<ExpirableToken>>>,
+    get_time_now_fn: Arc<dyn Fn() -> Duration + Send + Sync + 'static>,
+}
+
+#[derive(Clone)]
+struct BasicAuthExtension {
+    user: String,
+    password: SensitiveString,
+}
+
+#[derive(Debug, Deserialize)]
+struct Token {
+    access_token: String,
+    #[serde(rename = "expires_in")] 
```

Comment:
Ah sorry, was this a reserved name for an interface?

Reply:
this is the response from oauth2 server, so I need to remap it

Reply:
Ah I see, what you did makes sense 👍

Reply:
so, can I rename it back to expires_in ?

Reply:
I pushed revert to expires_in

Reply:
I read comment once again - I guess you wanted to put a comment about a time unit behind this property ? if yes, one is already on place 
```rust
const fn calculate_valid_until(now: Duration, grace_period: u32, token: &Token) -> u128 {
    // 'expires_in' means, in seconds, for how long it will be valid, lets say 5min,
    // to not cause some random 4xx, because token expired in the meantime, we will make some
    // room for token refreshing, this room is a grace_period.
    let (mut grace_period_seconds, overflow) = token.expires_in.overflowing_sub(grace_period);
```
I also added explicit comment here
```rust
#[derive(Debug, Deserialize)]
struct Token {
    access_token: String,
    // This property, according to RFC, is expected to be in seconds.
    expires_in: u32,
}
```

Reply:
This is good  now 👍